src/map/status.c

```diff
@@ -12278,38 +12279,118 @@ bool status_readdb_sizefix(char* fields[], int columns, int current)
-	random_bonus = atoi(fields[3]);
-
-	for(i = 0; i < MAX_REFINE; i++)
-	{
-		char* delim;
-
-		if (!(delim = strchr(fields[4+i], ':')))
-			return false;
-
-		*delim = '\0';
-
-		status->dbs->refine_info[current].chance[i] = atoi(fields[4+i]);
+/**
+ * Processes a refine_db.conf entry.
+ *
+ * @param *r	  Libconfig setting entry. It is expected to be valid and it
+ *                won't be freed (it is care of the caller to do so if
+ *                necessary)
+ * @param n       Ordinal number of the entry, to be displayed in case of
+ *                validation errors.
+ * @param *source Source of the entry (file name), to be displayed in case of
+ *                validation errors.
+ * @return # of the validated entry, or 0 in case of failure.
+ */
+int status_readdb_refine_libconfig_sub(config_setting_t *r, const char *name, const char *source) {
+	config_setting_t *rate = NULL;
+	int i=0, type=0, bonus_per_level=0, rnd_bonus_v=0, rnd_bonus_lv=0;
+	
+	if(!strncmp(name, "Armors", 6))
+		type = 0;
+	else if(!strspn(&name[strlen(name)-1], "0123456789") || !(type = atoi(strtok(strdup(name), "WeaponLevel")))) {
+		ShowError("status_readdb_refine_libconfig_sub: Invalid key name for entry '%s' in \"%s\", skipping.\n", name, source);
+		return 0;
+	}
+	
+	if(type < REFINE_TYPE_ARMOR || type >= REFINE_TYPE_MAX ) {
+		ShowError("status_readdb_refine_libconfig_sub: Out of range level for entry '%s' in \"%s\", skipping.\n", name, source);
+		return 0;
+	} else if( !libconfig->setting_lookup_int(r, "StatsPerLevel", &bonus_per_level) ) {
+		ShowWarning("status_readdb_refine_libconfig_sub: Missing StatsPerLevel for entry '%s' in \"%s\", skipping.\n", name, source);
+		return 0;
+	} else if( !libconfig->setting_lookup_int(r, "RandomBonusStartLevel", &rnd_bonus_lv) ) {
+		ShowWarning("status_readdb_refine_libconfig_sub: Missing RandomBonusStartLevel for entry '%s' in \"%s\", skipping.\n", name, source);
+		return 0;
+	} else if ( !libconfig->setting_lookup_int(r, "RandomBonusValue", &rnd_bonus_v) ) {
+		ShowWarning("status_readdb_refine_libconfig_sub: Missing RandomBonusValue for entry '%s' in \"%s\", skipping.\n", name, source);
+		return 0;
+	}
 
-		if (i >= random_bonus_start_level - 1)
-			status->dbs->refine_info[current].randombonus_max[i] = random_bonus * (i - random_bonus_start_level + 2);
+	if ((rate=libconfig->setting_get_member(r, "Rates")) && config_setting_is_group(rate)) {
+		config_setting_t *t = NULL;
+		bool duplicate[MAX_REFINE];
+		int bonus_[MAX_REFINE], rnd_bonus[MAX_REFINE], chance_[MAX_REFINE];
+
+		memset(&duplicate,0,sizeof(duplicate));
+		memset(&bonus_,0,sizeof(bonus_));
+		memset(&rnd_bonus,0,sizeof(rnd_bonus));
+		memset(&chance_,0,sizeof(chance_));
+		i=0;
+		while( (t = libconfig->setting_get_elem(rate,i++)) && config_setting_is_group(t) ) {
+			int level=0, chance=0, bonus=0;
+			char *rlvl = config_setting_name(t);
+			
+			if(!strspn(&rlvl[strlen(rlvl)-1], "0123456789") || (level = atoi(strtok(rlvl, "Lv"))) <= 0 ) {
```

Comment:
This doesn't need `strtok`. `strncmp` is what you're looking for.

By the way, from the `strtok` manual page:

> The `strtok()` function is used to isolate sequential tokens in a null-terminated string, `str`.  These tokens are separated in the string by **at least one of the characters in `sep`**.  [...]

In other words, `strtok(foo, "Lv")` uses as separator **either** `'L'` **or** `'v'`, not the `"Lv"` string.